samza-sql-shell/src/main/java/org/apache/samza/sql/client/cli/CliShell.java

```diff
@@ -0,0 +1,827 @@
+import org.jline.utils.AttributedStyle;
+import org.jline.utils.InfoCmp;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+
+class CliShell {
+    private final Terminal m_terminal;
+    private final PrintWriter m_writer;
+    private final LineReader m_lineReader;
+    private final String m_1stPrompt;
+    private final SqlExecutor m_executor;
+    private final ExecutionContext m_exeContext;
+    private CliEnvironment m_env;
+    private boolean m_keepRunning = true;
+    private Map<Integer, String> m_executions = new TreeMap<>();
+
+    public CliShell(SqlExecutor executor, CliEnvironment environment, ExecutionContext execContext) {
+        if(executor == null || environment == null || execContext == null) {
+            throw new IllegalArgumentException();
+        }
+
+        // Terminal
+        try {
+            m_terminal = TerminalBuilder.builder()
+                .name(CliConstants.WINDOW_TITLE)
+                .build();
+        }
+        catch(IOException e) {
+            throw new CliException("Error when creating terminal", e);
+        }
+
+        // Terminal writer
+        m_writer = m_terminal.writer();
+
+        // LineReader
+        final DefaultParser parser = new DefaultParser()
+            .eofOnEscapedNewLine(true)
+            .eofOnUnclosedQuote(true);
+        m_lineReader = LineReaderBuilder.builder()
+            .appName(CliConstants.APP_NAME)
+            .terminal(m_terminal)
+            .parser(parser)
+            .highlighter(new CliHighlighter())
+            .completer(new StringsCompleter(CliCommandType.getAllCommands()))
+            .build();
+
+        // Command Prompt
+        m_1stPrompt = new AttributedStringBuilder()
+            .style(AttributedStyle.DEFAULT.foreground(AttributedStyle.YELLOW))
+            .append(CliConstants.PROMPT_1ST + CliConstants.PROMPT_1ST_END)
+            .toAnsi();
+
+        // Execution context and executor
+        m_env = environment;
+        m_env.takeEffect();
+        m_exeContext = execContext;
+        m_executor = executor;
+        m_executor.start(m_exeContext);
+    }
+
+    Terminal getTerminal() {
+        return m_terminal;
+    }
+
+    CliEnvironment getEnvironment() {
+        return m_env;
+    }
+
+    SqlExecutor getExecutor() {
+        return m_executor;
+    }
+
+    ExecutionContext getExeContext() {
+        return m_exeContext;
+    }
+
+    /**
+     *  Actually run the shell. Does not return until user choose to exit.
+     */
+    public void open() {
+        // Remember we cannot enter alternate screen mode here as there is only one alternate
+        // screen and we need it to show streaming results. Clear the screen instead.
+        clearScreen();
+        m_writer.write(CliConstants.WELCOME_MESSAGE);
+
+        try {
+            // Check if jna.jar exists in class path
+            try {
+                ClassLoader.getSystemClassLoader().loadClass("com.sun.jna.NativeLibrary");
```

Comment:
Our shell uses jline which in turn uses jna.jar. If jna.jar is absent, the application will still be able to run but with a worse performance. This check is to prevent the performance loss from happening unnoticed.